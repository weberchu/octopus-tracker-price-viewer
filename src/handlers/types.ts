export type StandardUnitRates = {
    count: number,
    results: Result[],
}

export type Result = {
    value_exc_vat: number,
    value_inc_vat: number,
    valid_from: string,
    valid_to: string,
}