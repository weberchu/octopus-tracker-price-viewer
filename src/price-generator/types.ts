export type OctopusStandardUnitRates = {
    count: number,
    results: OctopusResult[],
}

export type OctopusResult = {
    value_exc_vat: number,
    value_inc_vat: number,
    valid_from: string,
    valid_to: string,
}

export enum Region {
    Eastern_England = "A",
    East_Midlands = "B",
    London = "C",
    Merseyside_and_Northern_Wales = "D",
    West_Midlands = "E",
    North_Eastern_England = "F",
    North_Western_England = "G",
    Southern_England = "H",
    South_Eastern_England = "J",
    Southern_Wales = "K",
    South_Western_England = "L",
    Yorkshire_= "M",
    Southern_Scotland = "N",
    Northern_Scotland = "P",
}

export type Price = {
    date: string
    electricityPrice?: string
    gasPrice?: string
}
export type Prices = {
    prices: Price[],
    lastUpdateTime: string,
}