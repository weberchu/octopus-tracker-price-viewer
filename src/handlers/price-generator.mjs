import AWS from 'aws-sdk';
const s3 = new AWS.S3();

const key = `price.html`;
const bucketName = 'octopus-tracker';

async function uploadToS3(fileContent) {
    const params = {
        Bucket: bucketName,
        Key: key,
        Body: fileContent,
    };

    try {
        const response = await s3.upload(params).promise();
        console.info('File uploaded successfully:', response.Location);
    } catch (error) {
        console.error('Error uploading file:', error);
        throw error;
    }
}

export const handler = async (event, context) => {
    console.info(JSON.stringify(event));

    const fileContent = 'Hello, S3!';
    return await uploadToS3(fileContent);
}
