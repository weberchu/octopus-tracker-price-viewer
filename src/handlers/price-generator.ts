import AWS from 'aws-sdk';
import fs from 'fs';
import Handlebars from "handlebars";
import fetch from 'node-fetch';
import { APIGatewayProxyEvent } from "aws-lambda";
import { Result, StandardUnitRates } from "./types";
import templateLocation from "../template/price.hbs";

type Price = {
    date: string
    electricityPrice?: string
    gasPrice?: string
}

type Prices = {
    prices: Price[],
    lastUpdateTime: string,
}

const s3 = new AWS.S3();
const key = `price.html`;
const bucketName = 'octopus-tracker';
const octopusApi = {
    baseProductUrl: "https://api.octopus.energy/v1/products/",
    productCode: "SILVER-FLEX-22-11-25",
    electricityTariffPrefix: "E-1R-",
    gasTariffPrefix: "G-1R-",
    regionCode: "C",
    path: {
        electricityTariffPath: "electricity-tariffs",
        gasTariffPath: "gas-tariffs",
        standardUnitRates: "standard-unit-rates"
    }
}

const electricityUrl =
    `${octopusApi.baseProductUrl}${octopusApi.productCode}` +
    `/${octopusApi.path.electricityTariffPath}` +
    `/${octopusApi.electricityTariffPrefix}${octopusApi.productCode}-${octopusApi.regionCode}` +
    `/${octopusApi.path.standardUnitRates}`;

const gasUrl =
    `${octopusApi.baseProductUrl}${octopusApi.productCode}` +
    `/${octopusApi.path.gasTariffPath}` +
    `/${octopusApi.gasTariffPrefix}${octopusApi.productCode}-${octopusApi.regionCode}` +
    `/${octopusApi.path.standardUnitRates}`;

const uploadToS3 = async (fileContent: string) => {
    const params = {
        Bucket: bucketName,
        Key: key,
        Body: fileContent,
        CacheControl: "max-age=10",
        ContentType: "text/html"
    };

    try {
        const response = await s3.upload(params).promise();
        console.info('File uploaded successfully:', response.Location);
    } catch (error) {
        console.error('Error uploading file:', error);
        throw error;
    }
}

const findPrice = (date: number, results: Result[]) => {
    return results.find((result) => {
        const from = Date.parse(result.valid_from);
        const to = Date.parse(result.valid_to);
        return from <= date && date <= to;
    })?.value_inc_vat;
}

const getDateString = (date: number) => {
    return new Date(date).toLocaleString('en-GB', {
        timeZone: "Europe/London",
        day: '2-digit',
        month: '2-digit'
    });
}

const getSingleFuelResult = async (url: string) => {
    const response = await fetch(url);
    const data = await response.json() as StandardUnitRates;
    return data.results;
}

export const getPrices = async () => {
    const electricityPrices = await getSingleFuelResult(electricityUrl);
    const gasPrices = await getSingleFuelResult(gasUrl);

    const today = Date.now()
    const yesterday = today - 24 * 60 * 60 * 1000;
    const tomorrow = today + 24 * 60 * 60 * 1000;

    const prices: Prices = {
        prices: [
            {
                date: getDateString(yesterday),
                electricityPrice: findPrice(yesterday, electricityPrices)?.toFixed(2),
                gasPrice: findPrice(yesterday, gasPrices)?.toFixed(2),
            },
            {
                date: getDateString(today),
                electricityPrice: findPrice(today, electricityPrices)?.toFixed(2),
                gasPrice: findPrice(today, gasPrices)?.toFixed(2),
            }
        ],
        lastUpdateTime: new Date().toLocaleString('en-GB', {
            timeZone: 'Europe/London',
            day: '2-digit',
            month: '2-digit',
            hour: '2-digit',
            minute: '2-digit'
        })
    };

    const tomorrowElectricityPrice = findPrice(tomorrow, electricityPrices)?.toFixed(2);
    const tomorrowGasPrice = findPrice(tomorrow, gasPrices)?.toFixed(2);
    if (tomorrowElectricityPrice || tomorrowGasPrice) {
        prices.prices.push({
            date: getDateString(tomorrow),
            electricityPrice: tomorrowElectricityPrice,
            gasPrice: tomorrowGasPrice,
        });
    }

    return prices;
}

const generateHtml = (prices: Prices) => {
    const template = fs.readFileSync(templateLocation, {encoding: "utf8", flag: "r"});
    const compiledTemplate = Handlebars.compile(template);
    return compiledTemplate(prices);
}

export const handler = async (event: APIGatewayProxyEvent) => {
    console.info("handler started");
    const prices = await getPrices();
    console.info("prices generated");
    const htmlContent = generateHtml(prices);
    console.info("htmlContent = " + htmlContent);
    return await uploadToS3(htmlContent);
}
