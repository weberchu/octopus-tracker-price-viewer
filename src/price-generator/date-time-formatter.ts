export const toDateString = (date: Date) => {
    return date.toLocaleString("en-GB", {
        timeZone: "Europe/London",
        day: "2-digit",
        month: "2-digit",
    });
};

export function toLastUpdateTime(date: Date) {
    return date.toLocaleString("en-GB", {
        timeZone: "Europe/London",
        day: "2-digit",
        month: "2-digit",
        hour: "2-digit",
        minute: "2-digit",
    });
}
