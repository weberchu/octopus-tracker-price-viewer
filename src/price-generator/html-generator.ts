import { Prices } from "./types";
import fs from "fs";
import templateLocation from "../template/price.hbs";
import Handlebars from "handlebars";

export const generateHtml = (prices: Prices) => {
    const template = fs.readFileSync(templateLocation, {encoding: "utf8", flag: "r"});
    const compiledTemplate = Handlebars.compile(template);
    return compiledTemplate(prices);
}