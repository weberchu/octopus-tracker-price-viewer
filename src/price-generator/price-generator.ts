import { Prices, Region } from "./types";
import { electricityUrl, gasUrl } from "./url-generator";
import { getSingleFuel } from "./single-fuel-fetcher";
import { findPrice } from "./price-finder";
import { toDateString, toLastUpdateTime } from "./date-time-formatter";

export const getPrices = async (): Promise<Prices> => {
    const electricityPrices = await getSingleFuel(electricityUrl(Region.London));
    const gasPrices = await getSingleFuel(gasUrl(Region.London));

    const now = new Date();
    const todayEpoch = now.getTime();
    const yesterday = new Date(todayEpoch - 24 * 60 * 60 * 1000);
    const tomorrow = new Date(todayEpoch + 24 * 60 * 60 * 1000);

    const prices: Prices = {
        prices: [
            {
                date: toDateString(yesterday),
                electricityPrice: findPrice(yesterday, electricityPrices)?.toFixed(2),
                gasPrice: findPrice(yesterday, gasPrices)?.toFixed(2),
            },
            {
                date: toDateString(now),
                electricityPrice: findPrice(now, electricityPrices)?.toFixed(2),
                gasPrice: findPrice(now, gasPrices)?.toFixed(2),
            }
        ],
        lastUpdateTime: toLastUpdateTime(now)
    };

    const tomorrowElectricityPrice = findPrice(tomorrow, electricityPrices)?.toFixed(2);
    const tomorrowGasPrice = findPrice(tomorrow, gasPrices)?.toFixed(2);
    if (tomorrowElectricityPrice || tomorrowGasPrice) {
        prices.prices.push({
            date: toDateString(tomorrow),
            electricityPrice: tomorrowElectricityPrice,
            gasPrice: tomorrowGasPrice,
        });
    }

    return prices;
}

