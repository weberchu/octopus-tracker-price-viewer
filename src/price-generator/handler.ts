import { APIGatewayProxyEvent } from "aws-lambda";
import { generateHtml } from "./html-generator";
import { uploader } from "./uploader";
import { getPrices } from "./price-generator";

export const handler = async (event: APIGatewayProxyEvent) => {
    console.info("handler started");
    const prices = await getPrices();
    console.info("prices generated");
    const htmlContent = generateHtml(prices);
    console.info("htmlContent = " + htmlContent);
    return await uploader.uploadToS3(htmlContent);
}