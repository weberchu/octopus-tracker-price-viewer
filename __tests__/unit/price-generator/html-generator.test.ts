import { generateHtml } from "../../../src/price-generator/html-generator";
import { Prices } from "../../../src/price-generator/types";

describe("html-generator", () => {
    describe("generateHtml", () => {
        test("should match snapshot", () => {
            const prices: Prices = {
                prices: [
                    {
                        date: "10/07",
                        electricityPrice: "1.0",
                        gasPrice: "2.0",
                    },
                    {
                        date: "11/07",
                        electricityPrice: "3.0",
                        gasPrice: "4.0",
                    },
                    {
                        date: "12/07",
                        electricityPrice: "5.0",
                    },
                ],
                lastUpdateTime: "12:34:56"
            }
            const html = generateHtml(prices);

            expect(html).toMatchSnapshot();
        });
    });
});