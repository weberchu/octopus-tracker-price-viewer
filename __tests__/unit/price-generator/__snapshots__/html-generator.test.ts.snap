// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html-generator generateHtml should match snapshot 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Tracker Price</title>
    <style>
        body {
            font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
            background-color: #080808;
            margin: 0;
            padding: 0;
        }

        table {
            margin: 10px auto;
            border-collapse: collapse;
        }

        th, td {
            padding: 0.4vw 3vw;
            text-align: center;
            font-size: 8vw;
            color: #F7F7F7;
        }

        .lastUpdateTime {
            font-size: 6vw;
            color: #878787;
            text-align: center;
        }

        @media screen and (min-width: 1200px) {
            th, td {
                padding: 4.8px 36px;
                font-size: 96px;
            }
            .lastUpdateTime {
                font-size: 48px;
            }
        }
    </style>
</head>
<body>
<table>
    <thead>
    <tr>
        <th></th>
        <th>‚ö°Ô∏è</th>
        <th>üî•</th>
    </tr>
    </thead>
    <tbody>
        <tr>
            <td>10/07</td>
            <td>
                    1.0p
            </td>
            <td>
                    2.0p
            </td>
        </tr>
        <tr>
            <td>11/07</td>
            <td>
                    3.0p
            </td>
            <td>
                    4.0p
            </td>
        </tr>
        <tr>
            <td>12/07</td>
            <td>
                    5.0p
            </td>
            <td>
            </td>
        </tr>
    </tbody>
</table>
<div class="lastUpdateTime">12:34:56</div>
</body>
</html>
"
`;
